app/helpers/samples_helper.rb

```diff
@@ -262,26 +260,54 @@ def sample_uploader(sample)
     user
   end
 
-  def sample_derived_data(sample)
+  def sample_derived_data(sample, job_stats_hash)
     output_data = {}
     pipeline_run = sample.pipeline_runs.first
-    job_stats = pipeline_run ? pipeline_run.job_stats : nil
-    summary_stats = job_stats ? get_summary_stats(job_stats) : nil
+    summary_stats = job_stats_hash.present? ? get_summary_stats(job_stats_hash, pipeline_run) : nil
     output_data[:pipeline_run] = pipeline_run
     output_data[:host_genome_name] = sample.host_genome ? sample.host_genome.name : nil
-    output_data[:job_stats] = job_stats
     output_data[:summary_stats] = summary_stats
 
     output_data
   end
 
+  def job_stats_db_multiget(pipeline_run_ids)
+    all_job_stats = JobStat.where(pipeline_run_id: pipeline_run_ids)
+    job_stats_by_pipeline_run_id = {}
+    all_job_stats.each do |entry|
+      job_stats_by_pipeline_run_id[entry.pipeline_run_id] ||= {}
+      job_stats_by_pipeline_run_id[entry.pipeline_run_id][entry.task] = entry
+    end
+    job_stats_by_pipeline_run_id
+  end
+
+  def job_stats_db_get(pipeline_run_id)
+    job_stats_db_multiget([pipeline_run_id])[pipeline_run_id]
+  end
+
+  def report_ready_db_multiget(pipeline_run_ids)
+    # get ids of pipeline_runs for which "report_ready?" is true
+    PipelineRun.where(id: pipeline_run_ids).where("job_status = '#{PipelineRun::STATUS_CHECKED}' or id in (select pipeline_run_id from pipeline_run_stages where pipeline_run_stages.step_number = pipeline_runs.ready_step and pipeline_run_stages.job_status = '#{PipelineRun::STATUS_LOADED}')").pluck(:id)
+  end
+
   def format_samples(samples)
     formatted_samples = []
+    return formatted_samples if samples.empty?
+
+    # Do major SQL queries
+    sample_ids = samples.map(&:id)
+    pipeline_run_ids = PipelineRun.where("id in (select max(id) from pipeline_runs where
```

Comment:
I think you can get full pipeline_runs here and you can contruct a sample_id to pipeline_run hash say pipeline_run_by_sample_id 
and then you can go pipeline_run_by_sample_id.values.map(&:id) to get list of pipeline_run_ids and reuse the pipeline_runs object below.